src/CacheManager.Core/Internal/CacheKeysHelper.cs

```diff
@@ -0,0 +1,115 @@
+        public static IEnumerable<string> FilterBy(this IEnumerable<string> keys, string pattern)
+        {
+            switch (MatchType(pattern))
+            {
+                case KeyMatchPattern.ALL:
+                    return keys;
+                case KeyMatchPattern.CONTAINS_STRING:
+                    return keys.Where(k => k.Contains(pattern));
+                case KeyMatchPattern.ENDS_WITH_CHAR:
+                    return keys.Where(k => k.Length == pattern.Length + 1 && k.StartsWith(pattern));
+                case KeyMatchPattern.ENDS_WITH_WILDCARD:
+                    return keys.Where(k => k.StartsWith(pattern));
+                case KeyMatchPattern.STARTS_WITH_CHAR:
+                    return keys.Where(k => k.Length == pattern.Length + 1 && k.EndsWith(pattern));
+                case KeyMatchPattern.STARTS_WITH_WILDCARD:
+                    return keys.Where(k => k.EndsWith(pattern));
+                case KeyMatchPattern.CONTAINS_SINGLE_CHAR:
+                    {
+                        var patterns = pattern.Split('?');
+                        return keys.Where(k => k.Length == pattern.Length + 1 && k.StartsWith(patterns[0]) && k.EndsWith(patterns[1]));
+                    }
+                case KeyMatchPattern.CONTAINS_SINGLE_WILDCARD:
+                    {
+                        var patterns = pattern.Split('?');
+                        return keys.Where(k => k.StartsWith(patterns[0]) && k.EndsWith(patterns[1]));
+                    }
+                default:
+                    {
+                        var regexPattern = "^" + Regex.Escape(pattern).Replace(@"\*", ".*").Replace(@"\?", ".") + "$";
+                        var regex = new Regex(regexPattern);
+                        return keys.Where(k => regex.IsMatch(k));
+                    }
+            }
+        }
+
+        static KeyMatchPattern MatchType(string pattern)
+        {
+            if (pattern == "*")
+                return KeyMatchPattern.ALL;
+
+            var firstCharMatch = pattern.IndexOf('?');
+
+            if (firstCharMatch == -1)
+            {
+                var firstWildcard = pattern.IndexOf('*');
+                if (firstWildcard == -1)
+                    return KeyMatchPattern.CONTAINS_STRING;
+
+                var lastWildcard = pattern.LastIndexOf('*');
+                if (firstWildcard == lastWildcard)
+                {
+                    if (firstWildcard == 0)
+                        return KeyMatchPattern.STARTS_WITH_WILDCARD;
+                    if (firstWildcard == pattern.Length - 1)
+                        return KeyMatchPattern.ENDS_WITH_WILDCARD;
+                    return KeyMatchPattern.CONTAINS_SINGLE_WILDCARD;
+                }
+            }
+            else
+            {
+                var lastCharMatch = pattern.LastIndexOf('?');
+                if (firstCharMatch == lastCharMatch)
+                {
+                    if (firstCharMatch == 0)
```

Comment:
well, it is more about consistency, code style conventions in this project, and I personally find it easier to read like that ;)